Slice with nested concats not applied

### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
df = pl.DataFrame({'a': [1,2,3]})
x = pl.concat([df.lazy(), (df+10).lazy()])
y = pl.concat([(df+20).lazy(), (df+30).lazy()])
z = [x.slice(0,3), y.slice(0,1)]
p = pl.concat(z).collect()
q = pl.concat(pl.collect_all(z))
#print(p.shape) # (8,1)
#print(q.shape) # (4,1)
print(f'{pl.__version__=}'); assert p.equals(q), (p.shape, q.shape) # p should be equal to q

```

```ipython
In [2]: p
Out[2]: 
shape: (8, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 3   │
│ 11  │
│ 12  │
│ 13  │
│ 21  │
│ 31  │
└─────┘

In [3]: q
Out[3]: 
shape: (4, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 3   │
│ 21  │
└─────┘

```

### Log output

```shell
async thread count: 32
blocking thread count: 512
UNION: union is run in parallel
UNION: `slice is set` union is run sequentially
UNION: `slice is set` union is run sequentially
pl.__version__='1.40.1'
Traceback (most recent call last):
  File "<string>", line 10, in <module>
AssertionError: ((8, 1), (4, 1))
```

### Issue description

seems that `x.slice` and `y.slice` were pushed down to each of the concatenated lazyframes that make them, respectively, which is incorrect. however this only happens on concat-collect, whereas collect-concat gives correct

### Expected behavior

`p == q`

### Installed versions

<details>

```
--------Version info---------
Polars:              1.40.1
Index type:          UInt32
Platform:            Linux-6.8.0-106-generic-x86_64-with-glibc2.39
Python:              3.11.4 (main, Jul  5 2023, 13:45:01) [GCC 11.2.0]
Runtime:             rt32

----Optional dependencies----
Azure CLI            <not installed>
adbc_driver_manager  <not installed>
altair               5.5.0
azure.identity       <not installed>
boto3                <not installed>
cloudpickle          2.1.0
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2026.3.0
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           3.9.2
numpy                1.26.4
openpyxl             <not installed>
pandas               2.0.0
polars_cloud         <not installed>
pyarrow              20.0.0
pydantic             2.10.6
pyiceberg            <not installed>
sqlalchemy           1.4.46
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

--- a/crates/polars-plan/src/plans/optimizer/flatten_union.rs
+++ b/crates/polars-plan/src/plans/optimizer/flatten_union.rs
@@ -3,6 +3,7 @@ use polars_core::error::PolarsResult;
 use polars_utils::arena::{Arena, Node};
 
 use super::OptimizationRule;
+use crate::dsl::UnionOptions;
 use crate::prelude::IR;
 
 pub struct FlattenUnionRule {}
@@ -26,7 +27,7 @@ impl OptimizationRule for FlattenUnionRule {
         match lp {
             Union { inputs, options }
                 if inputs.iter().any(|node| match lp_arena.get(*node) {
-                    Union { options, .. } => !options.flattened_by_opt,
+                    Union { options, .. } => !options.flattened_by_opt && options.slice.is_none(),
                     _ => false,
                 }) =>
             {

--- a/crates/polars-plan/src/plans/optimizer/flatten_union.rs
+++ b/crates/polars-plan/src/plans/optimizer/flatten_union.rs
@@ -3,7 +3,6 @@ use polars_core::error::PolarsResult;
 use polars_utils::arena::{Arena, Node};
 
 use super::OptimizationRule;
-use crate::dsl::UnionOptions;
 use crate::prelude::IR;
 
 pub struct FlattenUnionRule {}

--- a/py-polars/tests/unit/lazyframe/test_optimizations.py
+++ b/py-polars/tests/unit/lazyframe/test_optimizations.py
@@ -909,3 +909,12 @@ def test_slice_pushdown_expr_height_rules() -> None:
     assert ".slice(" not in plan
 
     assert_frame_equal(q.collect(), pl.DataFrame({"a": 3}))
+
+
+def test_forbid_flatten_sliced_union_27455() -> None:
+    df = pl.DataFrame({"a": [0, 0, 0, 0, 0]})
+    q1 = pl.concat([(df + 1).lazy(), (df + 10).lazy()])
+    q2 = pl.concat([(df + 100).lazy(), (df + 1000).lazy()])
+    q = pl.concat([q1.head(1), q2.head(1)])
+
+    assert_frame_equal(q.collect(), pl.DataFrame({"a": [1, 100]}))

